### ISSUE TITLE
do not register overlapping old DH key exchange messages

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5352]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)| @pitrou reported|
|-|-|
|Trac ID|trac#5352|
|Type|defect|
|Created|2011-11-01 20:31:22Z|
|Branch|https://github.com/twisted/twisted/tree/conch-legacy-dh-messages-5352|

This is tricky. Some legacy SSH messages (MSG_KEXDH_*) have overlapping values with newere ones (MSG_KEX_DH_GEX_*). t.conch.ssh.transport maps all of them by iterating over its own globals. Since the globals are a dict, iteration order is undefined; SSH transports can break nastily if the legacy messages happen to override the newer ones in the mapping, since the former don't have public handlers. Patch attached.

Attachments:

* [ssh_legacy_dh_msgs.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/806/806cd5051e7cba4dd8d1267b65a25f40c9c75278/102b08f6d45897627b342afb45464818881298ec.patch) (458 bytes) - added by antoine on 2011-11-01 20:31:50Z - 
* [ssh_legacy_dh_msgs2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/806/806cd5051e7cba4dd8d1267b65a25f40c9c75278/1220b9a9020c38e26594baad954ce774bdd6f266.patch) (641 bytes) - added by antoine on 2011-11-12 19:33:59Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5352 5352
type__defect defect
reporter__antoine antoine
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_conch_legacy_dh_messages_5352 branches/conch-legacy-dh-messages-5352
branch_author__z3p z3p
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1320179482000000 1320179482000000
changetime__1325359243000000 1325359243000000
version__None None
owner__ 
cc__z3p cc__antoine
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_3"></a>|@pitrou commented|
|-|-|

New patch adds a runtime check against regressions, in order to detect them easily.
(this was Jean-Paul's suggestion on IRC)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|z3p commented|
|-|-|

(In [33124]) Branching to 'conch-legacy-dh-messages-5352'

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|z3p set status to `closed`|
|-|-|

(In [33132]) Merge conch-legacy-dh-messages-5352: Do not register overlapping old DH key exchange messages

Author: antoine
Reviewer: z3p
Fixes: [#5352](https://github.com/twisted/twisted/issues/5352)

twisted.conch.ssh.transport.messages is created by iterating over globals().
Since globals is a dictionary, the ordering is undefined, so we can end up with
old message IDs in messages.  This patch checks for those old IDs and doesn't
insert them into the dictionary.  While not ideal, until exarkun writes some
code for iterating over named constants, we can live with it.
